core/modules/wsched/ChunkDisk.cc

```diff
@@ -147,63 +135,79 @@ bool ChunkDisk::_ready() {
+        memman::MemMan::Handle handle = _memMan->lock(tblVect, chunkId);
+        LOGS(_log,LOG_LVL_DEBUG, "handle=" << handle);
+        if (handle == 0) {
+            switch (errno) {
+            case ENOMEM:
+                setResourceStarved(true);
+                return false;
+            case ENOENT:
+                LOGS(_log, LOG_LVL_ERROR, "_memMgr->lock errno=ENOENT chunk not found tSeq=" << task->tSeq);
+                // Not sure if this is the best course of action, but it should just need one
+                // logic path. The query should fail from the missing tables
+                // and the czar must be able to handle that with appropriate retries.
+                handle = memman::MemMan::HandleType::ISEMPTY;
+                break;
+            default:
+                LOGS(_log, LOG_LVL_ERROR, "_memMgr->lock file system error tSeq=" << task->tSeq);
+                // Any error reading the file system is probably fatal for the worker.
+                throw std::bad_exception();
+                return false;
+            }
+        }
+        task->setMemHandle(handle);
+        setResourceStarved(false);
+        // Once the chunk has been granted, everything equal and below must go on pending.
+        // Otherwise there's a risk of a Task with lower or same chunkId getting in front
+        // of this one and needing the resources this Task has been promised.
+        _lastChunk = chunkId;
+    }
+    return true;
 }
 
 /// Return a Task that is ready to run, if available.
-wbase::Task::Ptr ChunkDisk::getTask() {
+wbase::Task::Ptr ChunkDisk::getTask(bool useFlexibleLock) {
     LOGS(_log, LOG_LVL_DEBUG, "ChunkDisk::getTask start");
     std::lock_guard<std::mutex> lock(_queueMutex);
-    if (!_ready()) {
+    if (!_ready(useFlexibleLock)) {
         LOGS(_log, LOG_LVL_DEBUG, "ChunkDisk denying task");
         return nullptr;
     }
     // Check the chunkId.
     auto task = _activeTasks.pop();
-    int chunkId = taskChunkId(*task);
-    LOGS(_log, LOG_LVL_DEBUG, "ChunkDisk getTask: current= " << _chunkState
+    int chunkId = task->getChunkId();
+    LOGS(_log, LOG_LVL_DEBUG, "ChunkDisk getTask: current=" << _lastChunk
          << " candidate=" << chunkId << " tSeq=" << task->tSeq);
-    _chunkState.addScan(chunkId);
-    registerInflight(task); // consider the task inflight as soon as it's off the queue
     return task;
-    // If next chunk is of a different chunk, only continue if current
-    // chunk has completed a scan already.
-
-    // FIXME: If time for chunk has expired, advance to next chunk
-    // Get the next chunk from the queue.
 }
 
-bool ChunkDisk::busy() const {
-    std::lock_guard<std::mutex> lock(_queueMutex);
-    return _busy();
+// TODO: DM-4943 add statistics.
+void ChunkDisk::setResourceStarved(bool starved) {
+    if (starved != _resourceStarved) {
+        _resourceStarved = starved;
+        LOGS(_log, LOG_LVL_DEBUG, "resourceStarved changed to " << _resourceStarved);
```

Comment:
Sending bool to stream may produce interesting results, I usually wrap it into `int(boolVal)`.